File: Common/Src/ChinesePuzzle/Game/GameControlNode.cpp
/**
 *  GameControlNode.cpp
 *  ChinesePuzzle
 *
 *  Created by Mathieu LEDRU on 01/11/11.
 *
 *  GPL License:
 *  Copyright (c) 2011, Mathieu LEDRU
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *  
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *  
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include "GameControlNode.h"
#include "CCNode.h"
#include "Game.h"

using namespace cocos2d;

GameControlNode::GameControlNode()
{
}

GameControlNode::~GameControlNode()
{
}

void GameControlNode::step(ccTime dt)
{
	
}

void GameControlNode::draw()
{
	
}

void GameControlNode::addNode(CCNode* node)
{
	nodes.insert(node);
}

void GameControlNode::removeNode(CCNode* node)
{
	nodes.erase(node);
}

void GameControlNode::updateNode(CCNode* node)
{
	
}

CCNode* GameControlNode::checkPoint(cocos2d::CCPoint p)
{
	for(NodeNodeSet::const_iterator it = nodes.begin(); it != nodes.end(); ++it)
	{
		CCRect rIt = (*it)->boundingBox();
		rIt.origin = (*it)->convertToWorldSpace(CCPointZero);
		
		if (CCRect::CCRectContainsPoint(rIt, p))
		{
			return (*it);
		}
	}
	
	return NULL;
}

CCNode* GameControlNode::checkPointNode(CCNode* c)
{
	for(NodeNodeSet::const_iterator it = nodes.begin(); it != nodes.end(); ++it)
	{
		if((*it) == c) continue;
		
		CCPoint p = c->getPosition();
		CCRect rIt = (*it)->boundingBox();
		
		if (CCRect::CCRectContainsPoint(rIt, p))
		{
			return (*it);
		}
	}
	
	return NULL;
}

CCNode* GameControlNode::checkRect(cocos2d::CCRect r, Filter filter)
{
	CCNode* cRes = NULL;
	float minDist = -1;
	
	for(NodeNodeSet::const_iterator it = nodes.begin(); it != nodes.end(); ++it)
	{
		if(!filter(*it)) continue;
		
		CCRect rIt = (*it)->boundingBox();
		rIt.origin = (*it)->convertToWorldSpace(CCPointZero);
		
		if (CCRect::CCRectIntersectsRect(rIt, r))
		{
			CCPoint vect = ccpAdd(r.origin, ccpNeg(rIt.origin));
			float dist = vect.x * vect.x + vect.y * vect.y;
			if(minDist == -1 || dist < minDist)
			{
				minDist = dist;
				cRes = (*it);
			}
		}
	}
	
	return cRes;
}

CCNode* GameControlNode::checkRectNode(CCNode* c, Filter filter)
{
	CCNode* cRes = NULL;
	float minDist = -1;
	
	for(NodeNodeSet::const_iterator it = nodes.begin(); it != nodes.end(); ++it)
	{
		if(!filter(*it)) continue;
		
		CCRect r = c->boundingBox();
		CCRect rIt = (*it)->boundingBox();
		
		if (CCRect::CCRectIntersectsRect(rIt, r))
		{
			CCPoint vect = ccpAdd(r.origin, ccpNeg(rIt.origin));
			float dist = vect.x * vect.x + vect.y * vect.y;
			if(minDist == -1 || dist < minDist)
			{
				minDist = dist;
				cRes = (*it);
			}
		}
	}
	
	return cRes;
}

File: Common/Bin/Data/js/Game/GameControlNode.js
/*
This file is part of the ChinesePuzzle package.

(c) Mathieu Ledru

For the full copyright and license information, please view the LICENSE
file that was distributed with this source code.
*/

cpz.GameControlNode = cpz.GameControl.extend({
  _nodes: [],
  step: function(dt) {},
  draw: function() {},
  addNode: function(node) {
    if (this._nodes.indexOf(node) === -1) {
      return this._nodes.push(node);
    }
  },
  removeNode: function(node) {
    var i, _i, _ref, _results;
    if (this._nodes.indexOf(node) > -1) {
      _results = [];
      for (i = _i = 0, _ref = this._nodes.length; 0 <= _ref ? _i <= _ref : _i >= _ref; i = 0 <= _ref ? ++_i : --_i) {
        if (this._nodes[i] === node) {
          _results.push(this._nodes.splice(i, 1));
        } else {
          _results.push(void 0);
        }
      }
      return _results;
    }
  },
  updateNode: function(node) {},
  checkPoint: function(point) {
    var n, rectNode, _i, _len, _ref;
    _ref = this._nodes;
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      n = _ref[_i];
      rectNode = n.getBoundingBox();
      rectNode._origin = n.convertToWorldSpace(cc.PointZero());
      if (cc.rectContainsPoint(rectNode, point)) {
        return n;
      }
    }
    return null;
  },
  checkPointNode: function(node) {
    var n, point, rectNode, _i, _len, _ref;
    _ref = this._nodes;
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      n = _ref[_i];
      if (n === node) {
        continue;
      }
      point = node.getPosition();
      rectNode = n.getBoundingBox();
      if (cc.rectContainsPoint(rectNode, point)) {
        return node;
      }
    }
    return null;
  },
  checkRect: function(rect, filter) {
    var dist, minDist, n, nodeRes, rectNode, vect, _i, _len, _ref;
    nodeRes = null;
    minDist = -1;
    _ref = this._nodes;
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      n = _ref[_i];
      if (!filter(n)) {
        continue;
      }
      rectNode = n.getBoundingBox();
      rectNode._origin = n.convertToWorldSpace(cc.PointZero());
      if (cc.rectIntersectsRect(rectNode, rect)) {
        vect = cc.pAdd(rect._origin, cc.pNeg(rectNode._origin));
        dist = vect.x * vect.x + vect.y * vect.y;
        if (minDist === -1 || dist < minDist) {
          minDist = dist;
          nodeRes = n;
        }
      }
    }
    return nodeRes;
  },
  checkRectNode: function(node, filter) {
    var dist, minDist, n, nodeRes, rect, rectNode, vect, _i, _len, _ref;
    nodeRes = null;
    minDist = -1;
    _ref = this._nodes;
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      n = _ref[_i];
      if (!filter(n)) {
        continue;
      }
      rect = node.getBoundingBox();
      rectNode = n.getBoundingBox();
      if (cc.rectIntersectsRect(rectNode, rect)) {
        vect = cc.pAdd(rect._origin, cc.pNeg(rectNode._origin));
        dist = vect.x * vect.x + vect.y * vect.y;
        if (minDist === -1 || dist < minDist) {
          minDist = dist;
          nodeRes = n;
        }
      }
    }
    return nodeRes;
  }
});
